src/electionguard/precompute_buffers.cpp

```diff
@@ -0,0 +1,243 @@
+            if (getInstance().populate_OK) {
+                unique_ptr<Triple> triple1;
+                unique_ptr<Triple> triple2;
+                unique_ptr<Quadruple> quad;
+
+                // generate two triples and a quadruple
+                triple1 = make_unique<Triple>(elgamalPublicKey);
+                TripleEntry *triple_entry1 = new TripleEntry();
+                triple_entry1->set(move(triple1));
+                getInstance().triple_queue.push(triple_entry1);
+               
+                triple2 = make_unique<Triple>(elgamalPublicKey);
+                TripleEntry *triple_entry2 = new TripleEntry();
+                triple_entry2->set(move(triple2));
+                getInstance().triple_queue.push(triple_entry2);
+
+                quad = make_unique<Quadruple>(elgamalPublicKey);
+                QuadrupleEntry *quad_entry = new QuadrupleEntry();
+                quad_entry->set(move(quad));
+                getInstance().quadruple_queue.push(quad_entry);
+                getInstance().queue_lock.unlock();
+            } else {
+                getInstance().queue_lock.unlock();
+                return;
+            }
+        } while (getInstance().quadruple_queue.size() < getInstance().max);
+    }
+
+    void PrecomputeBufferContext::stop_populate()
+    {
+        getInstance().queue_lock.lock();
+        getInstance().populate_OK = false;
+        getInstance().queue_lock.unlock();
+    }
+
+    uint32_t PrecomputeBufferContext::get_max_queue_size()
+    {
+        return getInstance().max;
+    }
+
+    uint32_t PrecomputeBufferContext::get_current_queue_size()
+    {
+        return getInstance().quadruple_queue.size();
+    }
+      
+    std::unique_ptr<TwoTriplesAndAQuadruple> PrecomputeBufferContext::getTwoTriplesAndAQuadruple()
+    {
+        unique_ptr<TwoTriplesAndAQuadruple> result = make_unique<TwoTriplesAndAQuadruple>();
+
+        // take a lock while we get the triples and a quadruple
+        getInstance().queue_lock.lock();
+
+        // make sure there are enough in the queues
+        if ((getInstance().triple_queue.size() >= 2) && (getInstance().quadruple_queue.size() >= 1)) {
+            TripleEntry *triple_entry1 = getInstance().triple_queue.front();
+            unique_ptr<Triple> triple1 = triple_entry1->get_triple();
+            delete triple_entry1;
+            getInstance().triple_queue.pop();
+           
+            TripleEntry *triple_entry2 = getInstance().triple_queue.front();
+            unique_ptr<Triple> triple2 = triple_entry2->get_triple();
+            delete triple_entry2;
+            getInstance().triple_queue.pop();
+
+            QuadrupleEntry *quadruple_entry = getInstance().quadruple_queue.front();
+            unique_ptr<Quadruple> quad = quadruple_entry->get_quadruple();
+            delete quadruple_entry;
+            getInstance().quadruple_queue.pop();
+            
+            result->set(move(triple1), move(triple2), move(quad));
```

Comment:
i probably would have just constructed the object here, but it looks like the api design is expected to return an initialized but empty object back if this control flow doesn't run, is that right?

Reply:
Yes you are understanding it correctly. I could have returned a null if the queues were empty I suppose.